new webModulesUrl to support custom hosting structures

FredKSchott opened the pull request:
Resolve follow ups to #271, #191, https://www.pika.dev/npm/snowpack/discuss/105
/cc @francislavoie @postspectacular @ssilvert @ruudud

This attempts to simplify the work that we've already done on the Babel plugin to support relative paths PLUS support more use-cases like a custom hosting layout that doesn't match the layout on disk. 

### Summary
- `importType` has been removed. The value of `webModulesUrl` now dictates the import type.
- `webModulesUrlPath` is now `webModulesUrl`:

### Example use-cases supported
  - `"/"` (Default): all imports translated to `/web_modules/PACKAGE_NAME`
  - `"/foo"`: all imports translated to `/foo/web_modules/PACKAGE_NAME`
  - `"./"`: all imports translated to relative path, assuming "web_modules" lives at "/web_modules"
    - Note: matches current `"importType: relative"` behavior
  - `"../"`: all imports translated to relative path, assuming "web_modules" lives at "/web_modules" and prefixed with an additional "../"
  - `"./foo"`: all imports translated to relative path, assuming "web_modules" lives at "/foo/web_modules"
  - `"https://static.foobar.com/"`: all imports translated to a full absolute URL on https://static.foobar.com.

This gives you more control over how we translate paths by just asking for a prefix and not trying to do any on-disk resolution ourselves on top of the resolution that Babel is already doing to give us a path from the source file to the current working directory "root".

New tests have been added, but I'd love some extra help testing, especially if the solution in #271 still didn't work for you.

francislavoie commented:
Interesting!

This config seems to do what I need, but it's not the config I would've expected to work:

```
    ["./plugins/snowpack-babel-plugin.js", {
      "webModulesUrl": "./",
      "webModulesDir": "../web_modules",
      "moduleResolution": "node"
    }],
```

I expected `../` to be the one I needed, but 🤷‍♂️ , I'm happy with having something functional! 😛

francislavoie commented:
So it looks like this no longer supports using a directory other than `web_modules` in the URL?

francislavoie commented:
I feel like doing it this way is more reasonable (allows overriding `web_modules` in the URL as well)

```diff
diff --git a/assets/babel-plugin.js b/assets/babel-plugin.js
index 2f9afdc..9cf41c4 100644
--- a/assets/babel-plugin.js
+++ b/assets/babel-plugin.js
@@ -41,9 +41,9 @@ function rewriteImport(importMap, imp, file, webModulesUrl, shouldResolveRelativ
     if (mappedImport.startsWith('http://') || mappedImport.startsWith('https://')) {
       return mappedImport;
     } else if (webModulesUrl.startsWith('http://') || webModulesUrl.startsWith('https://')) {
-      return webModulesUrl + path.posix.join('web_modules', mappedImport);
+      return path.posix.join(webModulesUrl, mappedImport);
     } else {
-      const joinedImport = path.posix.join(webModulesUrl, 'web_modules', mappedImport);
+      const joinedImport = path.posix.join(webModulesUrl, mappedImport);
       if (joinedImport.startsWith('/')) {
         return joinedImport;
       }
```

Then the default is `/web_modules`, I'll use `./web_modules` and `../foo/web_modules` is also possible.

francislavoie commented:
Also, does `//example.com/js/index.js` work for ES module imports in the browser? That's a common pattern for allowing the scheme to follow the page it's loaded in. I guess we'd just add `webModulesUrl.startsWith('//')` as a condition?

FredKSchott commented:
Good catch on "//", added.

Yea, I think you're right on "/" vs. "/web_modules". At the very least, it's a breaking change from the old "dir" option that we don't need to make right now. 

Added a note to the documentation that a relative import will be relative from the current working directory, or whatever you've set Babel's "root" as. I think it feels a bit more confusing as well, but it's better than the alternative of managing an additional config option and some other, competing concept of root.
